pilot/pkg/model/context.go

```diff
@@ -207,15 +215,40 @@ func ParseServiceNode(s string) (Proxy, error) {
 	default:
 		return out, fmt.Errorf("invalid node type (valid types: ingress, sidecar, router in the service node %q", s)
 	}
-	out.IPAddress = parts[1]
+
+	// Get all IP Addresses from Metadata
+	if ipstr, found := metadata["ISTIO_META_INSTANCE_IPS"]; found {
+		ipAddresses, err := parseIPAddresses(ipstr)
+		if err != nil {
+			out.IPAddresses = ipAddresses
+		} else if isValidIPAddress(parts[1]) {
+			//Fail back, use IP from node id
+			out.IPAddresses = append(out.IPAddresses, parts[1])
+		}
+	} else if isValidIPAddress(parts[1]) {
+		// Get IP from node id, it's only for backward-compatible, IP should come from metadata
+		out.IPAddresses = append(out.IPAddresses, parts[1])
+	}
 
 	// Does query from ingress or router have to carry valid IP address?
-	if net.ParseIP(out.IPAddress) == nil && out.Type == Sidecar {
-		return out, fmt.Errorf("invalid IP address %q in the service node %q", out.IPAddress, s)
+	if len(out.IPAddresses) == 0 && out.Type == Sidecar {
+		return out, fmt.Errorf("no valid IP address in the service node id or metadata")
 	}
 
 	out.ID = parts[2]
 	out.Domain = parts[3]
+
+	// Get all IP Addresses from Metadata
+	if ipstr, found := metadata["ISTIO_META_INSTANCE_IPS"]; found {
+		out.IPAddresses, _ = parseIPAddresses(ipstr)
```

Comment:
this whole logic seems strange. Why parse the IPaddresses twice?

Reply:
I wasn't aware of the second one. It may have been added accidentally when I copy & paste. It should be removed.

Reply:
I wasn't aware of the second one. It may have been added accidentally when I copy & paste. It should be removed.

Reply:
I wasn't aware of the second one. It may have been added accidentally when I copy & paste. It should be removed.